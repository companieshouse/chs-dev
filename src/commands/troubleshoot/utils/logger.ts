import { simpleColouriser } from "../../../helpers/colouriser.js";

export default class TroubleshootOutputLogger {

    private readonly logger: (msg: string) => void;

    constructor (logger: (msg: string) => void) {
        this.logger = logger;
    }

    logMessage (message: string, stat: "Success" | "Error"): void {
        stat === "Success"
            ? this.logger(`${simpleColouriser("SUCCESS!", "bold-green")} ${message}`)
            : this.logger(`${simpleColouriser("FAILED!", "red")} ${message}`);

    }
}
