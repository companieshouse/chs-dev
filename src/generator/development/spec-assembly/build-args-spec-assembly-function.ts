import Service from "../../../model/Service.js";
import { SpecAssemblyFunction } from "./spec-assembly-function.js";

type BuildArgApplicationFunction = (service: Service, composeServiceName: string, serviceSpec: {
    build?: Record<string, any>
}) => void

/**
 * Constructs a function which can handle the supplied build argument defined
 * by the argName parameter and will set its value when it is set on the input
 * spec
 * @param argName name of the build argument being set
 * @param valueSupplier function capable of supplying the value of the build
 *  arg from the service supplied
 * @returns function capable of setting args on the serviceSpec
 */
const buildArgApplicationFunctionFactory = (argName: string, valueSupplier: (service: Service) => string | undefined | null) => {
    return (service: Service, _: string, serviceSpec: {build?: Record<string, any>}) => {
        const value = valueSupplier(service);

        if (value !== null && typeof value !== "undefined" && typeof serviceSpec.build !== "undefined") {
            if (typeof serviceSpec.build.args === "undefined") {
                serviceSpec.build.args = {};
            }

            serviceSpec.build.args[argName] = value;
        }
    };
};

const buildArgApplicationFunctions: BuildArgApplicationFunction[] = [
    buildArgApplicationFunctionFactory("LANGUAGE_MAJOR_VERSION", service => service.metadata.languageMajorVersion),
    buildArgApplicationFunctionFactory("ENTRYPOINT", service => service.metadata.entrypoint),
    buildArgApplicationFunctionFactory("OUTDIR", service => service.metadata.buildOutputDir)
];

/**
 * A SpecAssemblyFunction for applying build args to the service specs within
 * the development docker compose spec file
 * @param developmentDockerComposeSpec spec file under construction
 * @param specAssemblyFunctionOptions SpecAssemblyFunctionOptions defining the
 *  options for assembling the spec
 */
const buildArgsSpecAssemblyFunction: SpecAssemblyFunction = (developmentDockerComposeSpec, {
    service
}) => {
    for (const serviceName of Object.keys(developmentDockerComposeSpec.services)) {
        buildArgApplicationFunctions.forEach(applicatorFunction => applicatorFunction(service, serviceName, developmentDockerComposeSpec.services[serviceName]));
    }
};

export default buildArgsSpecAssemblyFunction;
