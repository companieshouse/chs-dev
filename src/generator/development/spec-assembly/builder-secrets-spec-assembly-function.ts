import { getInitialDockerComposeFile } from "../../../helpers/docker-compose-file.js";
import CONSTANTS from "../../../model/Constants.js";
import { SpecAssemblyFunction } from "./spec-assembly-function.js";

const builderSecretsSpecAssemblyFunction: SpecAssemblyFunction = (
    developmentDockerComposeSpec,
    {
        projectPath,
        service
    }
) => {
    if (service.metadata.secretsRequired === CONSTANTS.BOOLEAN_LABEL_TRUE_VALUE) {
        const dockerComposeSpec = getInitialDockerComposeFile(projectPath);

        if (typeof dockerComposeSpec.secrets !== "undefined") {
            const serviceSpec = developmentDockerComposeSpec.services[`${service.name}-builder`] || developmentDockerComposeSpec.services[service.name];

            for (const secretName of Object.keys(dockerComposeSpec.secrets)) {
                if (typeof serviceSpec.secrets === "undefined") {
                    serviceSpec.secrets = [];
                }

                serviceSpec.secrets.push(
                    secretName
                );
            }
        }
    }
};

export default builderSecretsSpecAssemblyFunction;
