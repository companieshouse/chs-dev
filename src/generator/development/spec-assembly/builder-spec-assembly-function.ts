import { join } from "path";
import { SpecAssemblyFunction, SpecAssemblyFunctionOptions } from "./spec-assembly-function.js";
import yaml from "yaml";
import { DockerComposeSpec } from "../../../model/DockerComposeSpec.js";

const formatBuilderSpec = (builderSpec: string, serviceName: string, projectPath: string) => {
    const relativeRepositoryPath = join("repositories", serviceName);
    const replacements = [
        ["<service>", serviceName],
        ["<chs_dev_root>", projectPath],
        ["<repository_path>", relativeRepositoryPath],
        ["<absolute_repository_path>", join(projectPath, relativeRepositoryPath)]
    ];

    return replacements.reduce(
        (previous, [replacePattern, replaceValue]) => previous.replaceAll(replacePattern, replaceValue), builderSpec
    );
};

/**
 * SpecAssemblyFunction that applies the elements within the builder to the
 * development docker compose spec
 * @param developmentDockerComposeSpec spec under construction
 * @param specAssemblyFunctionOptions SpecAssemblyFunctionOptions defining the
 *  options for assembling the spec
 */
const builderSpecAssemblyFunction: SpecAssemblyFunction = (developmentDockerComposeSpec, {
    service,
    projectPath,
    builderDockerComposeSpec
}: SpecAssemblyFunctionOptions) => {
    // Apply the values in the builder spec now they are known and parse the yaml file
    const formattedBuilderSpec = formatBuilderSpec(builderDockerComposeSpec.builderSpec, service.name, projectPath);

    const builderSpec = yaml.parse(formattedBuilderSpec);

    // Collect any service which are not the service
    const newServices = Object.entries(builderSpec.services)
        .filter(([serviceName, serviceSpec]) => serviceName !== service.name);

    // Apply a dependency on these new service(s)
    for (const [newServiceName, newServiceSpec] of newServices) {
        developmentDockerComposeSpec.services[newServiceName] = newServiceSpec as DockerComposeSpec;

        if (typeof developmentDockerComposeSpec.services[service.name] !== "undefined") {
            if (typeof developmentDockerComposeSpec.services[service.name].depends_on === "undefined") {
                developmentDockerComposeSpec.services[service.name].depends_on = {
                    [newServiceName]: {
                        condition: "service_completed_successfully",
                        restart: true
                    }
                };
            } else {
                // @ts-expect-error - this is not undefined
                developmentDockerComposeSpec.services[service.name].depends_on[newServiceName] = {
                    condition: "service_completed_successfully",
                    restart: true
                };
            }
        }

    }

    // Merge the service attributes already in the development spec with the builders
    developmentDockerComposeSpec.services[service.name] = {
        ...builderSpec.services[service.name],
        ...developmentDockerComposeSpec.services[service.name]
    };

    // Repository or default behaviour has a few metadata items which can
    // customise the context as well as the build context
    if (builderDockerComposeSpec.name === "repository" &&
        typeof developmentDockerComposeSpec.services[service.name].build !== "undefined") {
        // append the repo context supplied in metadata to the existing value of context
        if (typeof service.metadata.repoContext !== "undefined") {
            // @ts-expect-error - this is not undefined
            const context = developmentDockerComposeSpec.services[service.name].build.context;
            // @ts-expect-error - this is not undefined
            developmentDockerComposeSpec.services[service.name].build.context = join(context, service.metadata.repoContext as string);
        }

        // Set the value of dockerfile to the value in metadata
        if (typeof service.metadata.dockerfile !== "undefined") {
            // @ts-expect-error - this is not undefined
            developmentDockerComposeSpec.services[service.name].build.dockerfile = service.metadata.dockerfile;
        }
    }
};

export default builderSpecAssemblyFunction;
