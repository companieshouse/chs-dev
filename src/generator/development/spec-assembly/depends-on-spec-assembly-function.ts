import { DockerComposeSpec } from "../../../model/DockerComposeSpec.js";
import { SpecAssemblyFunction } from "./spec-assembly-function.js";

const dependenciesAsObject = (dependencies: Record<string, any> | string[]) => {
    if (Array.isArray(dependencies)) {
        return dependencies.map(
            dependency => [dependency, { condition: "service_started" }]
        ).reduce((previous, [dependencyName, dependencySpec]) => ({
            ...previous,
            [dependencyName]: dependencySpec
        }), {});
    }

    return dependencies;
};

/**
 * Sets up the service's dependendencies correctly converting any services
 * using a list to a mapping
 * @param developmentDockerComposeSpec spec under construction
 * @param specAssemblyFunctionOptions SpecAssemblyFunctionOptions defining the
 *  options for assembling the spec
 */
const dependsOnSpecAssemblyFunction: SpecAssemblyFunction = (developmentDockerComposeSpec: DockerComposeSpec, {
    service,
    serviceDockerComposeSpec
}) => {
    if (typeof serviceDockerComposeSpec.services[service.name].depends_on !== "undefined") {
        developmentDockerComposeSpec.services[service.name].depends_on = {
            ...developmentDockerComposeSpec.services[service.name].depends_on,
            ...dependenciesAsObject(serviceDockerComposeSpec.services[service.name].depends_on as string[] | Record<string, any>)
        };
    }
};

export default dependsOnSpecAssemblyFunction;
