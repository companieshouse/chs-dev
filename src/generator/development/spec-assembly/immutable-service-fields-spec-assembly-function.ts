import { DockerComposeSpec } from "../../../model/DockerComposeSpec.js";
import { SpecAssemblyFunction } from "./spec-assembly-function.js";

const serviceSpecWithoutMutableFields = (serviceSpec: DockerComposeSpec, serviceName: string, mutableFields: string[]) => {
    return Object.entries(serviceSpec.services[serviceName])
        .filter(([serviceKey, _]) => !mutableFields.includes(serviceKey))
        .reduce((previous, [serviceKey, value]) => ({
            ...previous,
            [serviceKey]: value
        }), {});
};

/**
 * Creates a SpecAssemblyFunction which will apply any field present in the
 * service which are not part of the supplied list of mutable fields (i.e.
 * those fields mutated/modified by the spec assembly functions.)
 * @param mutableFields list of fields which may or may not be present in the
 *      source service which are modified by other SpecAssemblyFunctions
 * @returns SpecAssemblyFunction capable of copying across any field from the
 *      input service spec to the spec being generated which are not mutated
 */
export const immutableServiceFieldsSpecAssemblyFunction: (mutableFields: string[]) => SpecAssemblyFunction = (mutableFields: string[]) => (
    developmentDockerComposeSpec,
    {
        serviceDockerComposeSpec,
        service
    }
) => {
    developmentDockerComposeSpec.services[service.name] = serviceSpecWithoutMutableFields(
        serviceDockerComposeSpec,
        service.name,
        mutableFields
    );
};
