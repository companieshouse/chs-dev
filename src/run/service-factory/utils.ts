import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
