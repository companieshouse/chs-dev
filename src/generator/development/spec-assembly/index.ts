import { DockerComposeSpec } from "../../../model/DockerComposeSpec.js";
import buildArgsSpecAssemblyFunction from "./build-args-spec-assembly-function.js";
import builderSecretsSpecAssemblyFunction from "./builder-secrets-spec-assembly-function.js";
import builderSpecAssemblyFunction from "./builder-spec-assembly-function.js";
import dependsOnSpecAssemblyFunction from "./depends-on-spec-assembly-function.js";
import envFileSpecAssemblyFunction from "./env-file-assembly-function.js";
import { immutableServiceFieldsSpecAssemblyFunction } from "./immutable-service-fields-spec-assembly-function.js";
import nonServiceTlaSpecAssemblyFunction from "./non-services-tla-spec-assembly-function.js";
import { SpecAssemblyFunction, SpecAssemblyFunctionOptions } from "./spec-assembly-function.js";
import { volumeSpecAssemblyFunction } from "./volume-spec-assembly-function.js";

const specFieldsWhichAreMutable = [
    "build",
    "develop",
    "env_file",
    "volumes",
    "depends_on",
    "secrets"
];

const specAssemblyFunctions: SpecAssemblyFunction[] = [
    immutableServiceFieldsSpecAssemblyFunction(specFieldsWhichAreMutable),
    builderSpecAssemblyFunction,
    dependsOnSpecAssemblyFunction,
    buildArgsSpecAssemblyFunction,
    envFileSpecAssemblyFunction,
    nonServiceTlaSpecAssemblyFunction,
    volumeSpecAssemblyFunction,
    builderSecretsSpecAssemblyFunction
];

/**
 * Executes all the specAssemblyFunctions with the docker compose spec supplied
 * and supplied options
 * @param developmentDockerComposeSpec spec under construction
 * @param options SpecAssemblyFunctionOptions defining the
 *  options for assembling the spec
 */
const apply = (developmentDockerComposeSpec: DockerComposeSpec, options: SpecAssemblyFunctionOptions) => {
    specAssemblyFunctions.forEach(assemblyFunction => {
        assemblyFunction(developmentDockerComposeSpec, options);
    });
};

export default apply;
