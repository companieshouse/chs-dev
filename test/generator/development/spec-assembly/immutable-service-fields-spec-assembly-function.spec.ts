import { expect } from "@jest/globals";
import { createCompleteServiceSpecForServiceWithName } from "../../../utils/docker-compose-spec";
import { immutableServiceFieldsSpecAssemblyFunction } from "../../../../src/generator/development/spec-assembly/immutable-service-fields-spec-assembly-function";
import { services } from "../../../utils/data";

describe("immutableServiceFieldsSpecAssemblyFunction", () => {

    it("applies all attributes from service which are not in the mutable fields list", () => {
        const service = services[5];
        const serviceName = service.name;

        const serviceSpec = createCompleteServiceSpecForServiceWithName(serviceName);
        const developmentDockerComposeSpec = {
            services: {
                [serviceName]: {}
            }
        };

        const mutableFields = [
            "build",
            "develop",
            "env_file",
            "volumes",
            "depends_on",
            "secrets"
        ];

        const specAssemblyFunction = immutableServiceFieldsSpecAssemblyFunction(
            mutableFields
        );
        specAssemblyFunction(
            developmentDockerComposeSpec,
            {
                serviceDockerComposeSpec: serviceSpec,
                service,
                projectPath: "/user/test-user",
                builderDockerComposeSpec: {
                    name: "java",
                    version: "v1",
                    builderSpec: ""
                }
            }
        );

        expect(developmentDockerComposeSpec.services[serviceName]).toMatchSnapshot();
    });
});
