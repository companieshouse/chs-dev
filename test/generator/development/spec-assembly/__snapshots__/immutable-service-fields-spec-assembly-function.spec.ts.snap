// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`immutableServiceFieldsSpecAssemblyFunction applies all attributes from service which are not in the mutable fields list 1`] = `
{
  "annotations": [
    "one:two",
  ],
  "attach": false,
  "blkio_config": {
    "weight": 400,
  },
  "cap_add": [
    "ALL",
  ],
  "cap_drop": [
    "NET_ADMIN",
  ],
  "cgroup": "private",
  "cgroup_parent": "ingress",
  "command": "which java",
  "configs": [
    "config-one",
  ],
  "container_name": "my-container",
  "cpu_count": 1,
  "cpu_percent": 20,
  "cpu_period": 34,
  "cpu_quota": 1,
  "cpu_rt_period": "400ms",
  "cpu_rt_runtime": "400ms",
  "cpu_shares": 1,
  "cpus": 0.123,
  "cpuset": 3,
  "credential_spec": {
    "file": "my-cred-spec.json",
  },
  "deploy": {},
  "device_cgroup_rules": [
    "c 1:3 mr",
  ],
  "devices": [
    "/dev/ttyUSB0:/dev/ttyUSB0",
  ],
  "dns": [
    "8.8.8.8",
  ],
  "dns_opt": [
    "use-vpc",
  ],
  "dns_search": [
    "dc.example.com",
  ],
  "domainname": "hellworld.com",
  "driver_opts": {
    "optOne": "123456",
  },
  "entrypoint": "echo",
  "environment": {
    "ONE": "FOO",
    "TWO": "BAR",
  },
  "expose": 8998,
  "external_links": [
    "redis",
  ],
  "extra_hosts": [
    "hello.local=234.359.4567",
  ],
  "group_add": [
    "mail",
  ],
  "healthcheck": {
    "test": "echo 1",
  },
  "hostname": "hello.local",
  "image": "123456789.ecr.eu-west-2.com/service-six",
  "init": true,
  "ipc": "shareable",
  "isolation": "fp",
  "labels": [
    "label.one=foo",
    "label.two=bar",
  ],
  "links": [
    "db",
  ],
  "logging": {
    "driver": "syslog",
  },
  "mem_limit": "1024",
  "mem_reservation": "2048",
  "mem_swappiness": 50,
  "memswap_limit": "300m",
  "network_mode": "host",
  "networks": [
    "network-one",
  ],
  "oom_kill_disable": true,
  "oom_score_adj": 400,
  "pid": 2345,
  "pids_limit": 20,
  "platform": "darwin",
  "ports": [
    "9090:9090",
  ],
  "privileged": true,
  "profiles": [
    "frontend",
  ],
  "pull_policy": "missing",
  "read_only": true,
  "restart": "always",
  "runtime": "another",
  "scale": {},
  "security_opt": [
    "label:user:USER",
  ],
  "shm_size": "3094b",
  "stdin_open": false,
  "stop_grace_period": "1m30s",
  "stop_signal": "SIGUSR1",
  "storage_opt": {
    "size": "1G",
  },
  "sysctls": {
    "net.core.somaxconn": 1024,
  },
  "tmpfs": "/run",
  "tty": false,
  "ulimits": {
    "nproc": 65535,
  },
  "user": "fred-flintstone",
  "userns_mode": "host",
  "uts": "host",
  "volumes_from": [
    "service-for",
  ],
  "working_dir": "/opt/app/one-two-three",
}
`;
